src/goto-instrument/synthesizer/synthesizer_utils.cpp

```diff
@@ -0,0 +1,113 @@
+/*******************************************************************\
+
+Module: Utility functions for loop invariant synthesizer.
+
+Author: Qinheping Hu
+
+\*******************************************************************/
+
+#include "synthesizer_utils.h"
+
+goto_programt::const_targett get_loop_end_from_loop_head_and_content(
+  const goto_programt::const_targett &loop_head,
+  const loop_templatet<goto_programt::const_targett> &loop)
+{
+  goto_programt::const_targett loop_end = loop_head;
+  for(const auto &t : loop)
+  {
+    // an instruction is the loop end if it is a goto instruction
+    // and it jumps backward to the loop head
+    if(
+      t->is_goto() && t->get_target() == loop_head &&
+      t->location_number > loop_end->location_number)
+      loop_end = t;
+  }
+  INVARIANT(
+    loop_head != loop_end,
+    "Could not find end of the loop starting at: " +
+      loop_head->source_location().as_string());
+
+  return loop_end;
+}
+
+goto_programt::targett get_loop_end_from_loop_head_and_content_mutable(
+  const goto_programt::targett &loop_head,
+  const loop_templatet<goto_programt::targett> &loop)
+{
+  goto_programt::targett loop_end = loop_head;
+  for(const auto &t : loop)
+  {
+    // an instruction is the loop end if it is a goto instruction
+    // and it jumps backward to the loop head
+    if(
+      t->is_goto() && t->get_target() == loop_head &&
+      t->location_number > loop_end->location_number)
+      loop_end = t;
+  }
+  INVARIANT(
+    loop_head != loop_end,
+    "Could not find end of the loop starting at: " +
+      loop_head->source_location().as_string());
+
+  return loop_end;
+}
+
+goto_programt::targett get_loop_head_or_end(
+  const unsigned int target_loop_number,
+  goto_functiont &function,
+  bool finding_head)
+{
+  natural_loops_mutablet natural_loops(function.body);
+
+  // iterate over all natural loops to find the loop with `target_loop_number`
+  for(const auto &loop_p : natural_loops.loop_map)
```

Comment:
Is even this use of `natural_loops` necessary anymore? Can't you just iterate over all instructions and check the `loop_number` (you might want to use `is_backwards_goto()` to check that something is a potential loop-end).

Reply:
```natual_loops``` is necessary here. When we extract counterexample from trace, the trace is based on the goto model with loop contracts applied. However, in ```apply_loop_contracts```, we add backward goto instructions (for loop-contracts simplification) into the goto model. So we cannot use ```is_backwards_goto()``` to check if some instruction is a loop-end in a goto model with loop contracts applied.

Reply:
Just to add to @Herbping's reply, the reason we only `apply_loop_contracts` to `natural_loops` is because the instrumentation is unsound on loops that are not natural. For example, if a loop has jumps into its body from outside, then we could entirely skip over the initial base case assertion, havocing, invariant assumption etc.